File: python/2nd_ex/Pocklington.py
import random
from common import brute_force_of_divisors, rem_a

def pocklington_test(F, n, t):
    q = brute_force_of_divisors(F)
    a = [random.randint(2, n - 1) for _ in range(t)]
    if rem_a(a, n) == -1:
            return -1
    for i in a:
        res = i % n
        for j in q:
            deg = (n - 1) / j
            for _ in range(2, int(deg) + 1):
                res *= i
                res %= n
            if res == 1:
                break
        if res != 1:
            return 1
    return 0


def make_n(F, num_size):
    R = random.getrandbits(num_size // 2) 
    R |= 1 << (num_size // 2 - 1)
    R -= R & 1
    return R * F + 1

File: C++/2nd_ex/Pocklington.cpp
#include "common.h"

int pocklington_test(int F, int n, int t) {
  vector<int> q = brute_force_of_divisors(F);
  vector<int> a = {};
  int res;
  int deg;

  for (int i = 0; i < t; i++)
    a.push_back(gen_rand_num(2, n - 1));

  if (rem_a(a, n) == -1)
    return -1;

  for (auto i : a) {
    res = i % n;
    for (auto j : q) {
      deg = (n - 1) / j;
      for (int k = 2; k <= deg; k++)
        res = res * i % n;
      if (res == 1)
        break;
    }
    if (res != 1)
      return 1;
  }
  return 0;
}

int make_n(int F, int num_size) {
  int R = gen_rand_num(pow(2, num_size - 1), pow(2, num_size) - 1);
  R -= R & 1;
  return R * F + 1;
}
